src/lib.rs

```diff
@@ -1,215 +1,64 @@
-extern crate futures;
-extern crate parking_lot;
-
-use parking_lot::Mutex;
-use futures::prelude::*;
-use futures::task::{self, Task, AtomicTask};
-use futures::executor::{self, Notify};
-
-use std::collections::HashMap;
 use std::sync::Arc;
-use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::pin::Pin;
+use std::task::{Poll, Context};
+use std::ops::DerefMut;
+use futures::{Future, FutureExt, channel::oneshot, future::{select, Either}, executor::block_on};
+use parking_lot::Mutex;
 
-/// Future that resolves when inner work finishes or on exit signal firing.
+/// Future that resolves when the exit signal has fired.
 #[derive(Clone)]
-pub struct UntilExit<F> {
-    inner: F,
-    exit: Exit,
-}
-
-impl<F: Future> Future for UntilExit<F> {
-    type Item = Option<F::Item>;
-    type Error = F::Error;
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        match self.inner.poll() {
-            Ok(Async::Ready(x)) => Ok(Async::Ready(Some(x))),
-            Ok(Async::NotReady) => Ok(self.exit.check().map(|()| None)),
-            Err(e) => Err(e),
-        }
-    }
-}
-
-struct Notifier {
-    signalled: AtomicBool,
-    outer_task: AtomicTask,
-}
+pub struct Exit(Arc<Mutex<oneshot::Receiver<()>>>);
 
-impl Notify for Notifier {
-    // id isn't necessary because this is a notifier for a single
-    // exit future.
-    fn notify(&self, _: usize) {
-        self.signalled.store(true, Ordering::Release);
-        self.outer_task.notify()
-    }
-}
-
-struct ExitInner {
-    shared_id: usize,
-    notifier: Arc<Notifier>,
-}
-
-impl ExitInner {
-    fn check(&self, shared: &Shared) -> Async<()> {
-        // set up the outer task correctly so when the inner
-        // task is notified, we get woken up.
-        self.notifier.outer_task.register();
-
-        // if any poll was signalled by the inner task, use it.
-        if !self.notifier.signalled.fetch_and(false, Ordering::SeqCst) {
-            return Async::NotReady
-        }
+impl Future for Exit {
+    type Output = ();
 
-        // do heavy check with inner task. ID doesn't matter.
-        executor::with_notify(&self.notifier, 0, || {
-            if shared.is_live_and_notify(self.shared_id) {
-                Async::NotReady
-            } else {
-                Async::Ready(())
-            }
-        })
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut receiver = Pin::into_inner(self).0.lock();
+        Pin::new(receiver.deref_mut()).poll(cx).map(|_| ())
```

Comment:
This may `poll` after `Ready`, which I believe is fine to panic, which I think we should avoid :)

Reply:
Fixed